File: MemoryProxy/scripts/query_usage_stats.py
#!/usr/bin/env python3
"""Parse lightweight MemoryProxy JSONL logs and print usage/timing stats.

This is the lightweight alternative to ClickHouse for low-volume deployments:

  - PROXY_LOG_FILE/proxy.log        JSONL from FileLogger; rows have:
                                    {"timestamp":..., "level":..., "event":"request.timing", "data":{...}}
  - PROXY_LOG_FILE/YYYY-MM-DD.jsonl Daily JSONL from logger.ts; rows are
                                    {"timestamp":..., "event":"usage", "usage":{...}, "modelId":..., ...}

Old proxy.log text lines (before this JSONL change) are also handled by regex,
so it is safe to run this script against both old and new logs.

Usage examples:
  python3 scripts/query_usage_stats.py /data/tdai-memory-proxy/logs
  python3 scripts/query_usage_stats.py proxy.log 2026-09-07.jsonl
  python3 scripts/query_usage_stats.py --verbose /data/tdai-memory-proxy/logs
"""

import argparse
import json
import re
from pathlib import Path
from statistics import median

# Old line format: [2026-09-07T...][INFO] request.timing {...}
OLD_LOG_RE = re.compile(r"^\[[^\]]*\]\[[A-Z]+\]\s+(\S+)\s*(\{.*\})\s*$")


def parse_json_or_old_line(raw: str):
    """Return (record, event, data_dict) or (None, None, None)."""
    line = raw.strip()
    if not line:
        return None, None, None

    try:
        obj = json.loads(line)
    except json.JSONDecodeError:
        m = OLD_LOG_RE.match(line)
        if not m:
            return None, None, None
        event = m.group(1)
        try:
            data = json.loads(m.group(2))
        except json.JSONDecodeError:
            data = {}
        return None, event, data

    if not isinstance(obj, dict):
        return obj, None, None

    event = obj.get("event")
    if isinstance(event, str):
        data = obj.get("data")
        return obj, event, data if isinstance(data, dict) else {}
    return obj, None, obj


def parse_usage_tokens(usage) -> dict:
    if not isinstance(usage, dict):
        return {}
    try:
        prompt = usage.get("prompt_tokens")
        completion = usage.get("completion_tokens") or usage.get("output_tokens")
        cache_hit = usage.get("prompt_cache_hit_tokens") or usage.get("cache_read_input_tokens") or 0
        cache_write = usage.get("prompt_cache_write_tokens") or usage.get("cache_creation_input_tokens") or 0
        input_tokens = usage.get("input_tokens") or 0

        if prompt is None and input_tokens is not None:
            prompt = int(input_tokens or 0) + int(cache_hit or 0) + int(cache_write or 0)
        prompt = int(prompt or 0)
        completion = int(completion or 0)
        total = int(usage.get("total_tokens") or 0) or prompt + completion
        return {
            "prompt_tokens": prompt,
            "completion_tokens": completion,
            "total_tokens": total,
        }
    except (TypeError, ValueError):
        return {}


def percentile(values, q):
    if not values:
        return 0
    s = sorted(values)
    if len(s) == 1:
        return s[0]
    pos = q * (len(s) - 1)
    lo = int(pos)
    hi = min(lo + 1, len(s) - 1)
    return round(s[lo] + (s[hi] - s[lo]) * (pos - lo))


def collect_files(paths):
    files = []
    for p in paths:
        path = Path(p)
        if path.is_dir():
            for candidate in sorted(path.iterdir()):
                if candidate.is_file():
                    name = candidate.name
                    if name.endswith(".log") or name.endswith(".jsonl") or ".log." in name or ".jsonl." in name:
                        files.append(candidate)
        elif path.is_file():
            files.append(path)
    return files


def summarize_usage(records):
    rows = []
    for obj, event, data in records:
        if event not in ("usage", "analyzer_usage"):
            continue
        usage = obj.get("usage") if isinstance(obj, dict) else None
        if not isinstance(usage, dict):
            continue
        tokens = parse_usage_tokens(usage)
        if not tokens:
            continue
        rows.append({
            "timestamp": obj.get("timestamp") or data.get("timestamp") or "",
            "event": event,
            "model": obj.get("modelId") or data.get("modelId") or "unknown",
            "key_id": obj.get("keyId") or data.get("keyId") or "",
            "session_key": obj.get("sessionKey") or data.get("sessionKey") or "",
            **tokens,
        })
    return rows


def summarize_timing(records):
    rows = []
    for obj, event, data in records:
        if event == "request.timing":
            # New JSONL format stores fields under data; also accept top-level
            # fields if a future logger flattens them.
            if data:
                base = data
            else:
                base = obj if isinstance(obj, dict) else {}
            total = base.get("totalMs")
            if total is None:
                total = base.get("total_ms")
            if total is None:
                continue
            rows.append({
                "timestamp": base.get("timestamp") or obj.get("timestamp") if isinstance(obj, dict) else "",
                "model": base.get("model") or "unknown",
                "protocol": base.get("protocol") or "unknown",
                "stream": base.get("stream"),
                "total_ms": int(total or 0),
                "ttfb_ms": int(base.get("upstreamTtfbMs") or base.get("upstream_ttfb_ms") or 0),
                "stream_ms": int(base.get("upstreamStreamMs") or base.get("upstream_stream_ms") or 0),
            })
    return rows


def print_usage_stats(rows):
    print("\n== Token / Usage ==")
    if not rows:
        print("(no usage rows found)")
        return
    total_tokens = sum(r["total_tokens"] for r in rows)
    print(f"rows={len(rows)} total_tokens={total_tokens}")

    by_model = {}
    for r in rows:
        m = by_model.setdefault(r["model"], {"rows": 0, "total_tokens": 0})
        m["rows"] += 1
        m["total_tokens"] += r["total_tokens"]

    print(f"{'model':<20} {'calls':>6} {'tokens':>12}")
    for model, m in sorted(by_model.items(), key=lambda kv: -kv[1]["rows"]):
        print(f"{model:<20} {m['rows']:>6} {m['total_tokens']:>12}")


def print_timing_stats(rows):
    print("\n== Timing ==")
    if not rows:
        print("(no request.timing rows found)")
        return
    print(f"rows={len(rows)}")

    by_key = {}
    for r in rows:
        key = (r["model"], r["protocol"], r["stream"])
        by_key.setdefault(key, []).append(r)

    print(f"{'model':<16} {'protocol':<10} {'stream':>3} {'requests':>7} {'p50_total_ms':>13} {'p90_total_ms':>13} {'p50_ttfb_ms':>12} {'p90_ttfb_ms':>12}")
    for key, group in sorted(by_key.items(), key=lambda kv: -len(kv[1])):
        model, protocol, stream = key
        totals = [r["total_ms"] for r in group]
        ttfb = [r["ttfb_ms"] for r in group]
        print(
            f"{model:<16} {protocol:<10} {str(stream):>3} {len(group):>7} "
            f"{percentile(totals, 0.5):>13} {percentile(totals, 0.9):>13} "
            f"{percentile(ttfb, 0.5):>12} {percentile(ttfb, 0.9):>12}"
        )


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("paths", nargs="+", help="log file(s) or directory containing proxy.log / *.jsonl")
    parser.add_argument("--verbose", action="store_true", help="print malformed lines while scanning")
    args = parser.parse_args()

    files = collect_files(args.paths)
    if not files:
        print("No log files found in the given paths.")
        return 1

    records = []
    for path in files:
        try:
            with open(path, "r", encoding="utf-8", errors="replace") as fh:
                for line_no, raw in enumerate(fh, 1):
                    obj, event, data = parse_json_or_old_line(raw)
                    if event is None:
                        if args.verbose and raw.strip():
                            print(f"[skip] {path}:{line_no} not parseable")
                        continue
                    records.append((obj, event, data))
        except OSError as exc:
            print(f"[warning] cannot read {path}: {exc}")

    print_usage_stats(summarize_usage(records))
    print_timing_stats(summarize_timing(records))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: MemoryProxy/scripts/query_usage_stats.mjs
#!/usr/bin/env node
/**
 * Lightweight MemoryProxy JSONL usage/timing statistics parser.
 *
 * Run inside the proxy container or on any machine with Node.js:
 *
 *   node scripts/query_usage_stats.mjs /data/tdai-memory-proxy/logs
 *   node scripts/query_usage_stats.mjs proxy.log 2026-09-07.jsonl
 *
 * The parser understands:
 *   - New FileLogger JSONL lines:
 *       {"timestamp":"...","level":"INFO","event":"request.timing","data":{...}}
 *   - Daily usage JSONL from logger.ts:
 *       {"timestamp":"...","event":"usage","usage":{...},"modelId":"..."}
 *   - Old proxy.log prefix-text lines:
 *       [2026-09-07T...][INFO] request.timing {...}
 */

import fs from "node:fs";
import path from "node:path";

const OLD_LOG_RE = /^\[[^\]]*\]\[[A-Z]+\]\s+(\S+)\s*(\{.*\})\s*$/;

function collectFiles(inputPaths) {
  const files = [];
  for (const p of inputPaths) {
    const st = fs.statSync(p, { throwIfNoEntry: false });
    if (!st) continue;
    if (st.isDirectory()) {
      for (const name of fs.readdirSync(p).sort()) {
        const full = path.join(p, name);
        const child = fs.statSync(full);
        if (child.isFile() && (name.endsWith(".log") || name.endsWith(".jsonl") || name.includes(".log.") || name.includes(".jsonl."))) {
          files.push(full);
        }
      }
    } else if (st.isFile()) {
      files.push(p);
    }
  }
  return files;
}

function parseLine(raw) {
  const line = raw.trim();
  if (!line) return { event: null, data: null };

  try {
    const obj = JSON.parse(line);
    if (!obj || typeof obj !== "object" || Array.isArray(obj)) return { event: null, data: null };
    if (typeof obj.event === "string") {
      const data = obj.data && typeof obj.data === "object" ? obj.data : {};
      return { event: obj.event, data };
    }
    return { event: null, data: obj };
  } catch {
    const m = OLD_LOG_RE.exec(line);
    if (!m) return { event: null, data: null };
    try {
      return { event: m[1], data: JSON.parse(m[2]) };
    } catch {
      return { event: null, data: null };
    }
  }
}

function extractTokens(usage) {
  if (!usage || typeof usage !== "object") return null;
  const n = (v) => (typeof v === "number" && Number.isFinite(v) ? v : 0);
  const prompt = usage.prompt_tokens ?? (n(usage.input_tokens) + n(usage.prompt_cache_hit_tokens) + n(usage.prompt_cache_write_tokens));
  const completion = usage.completion_tokens ?? usage.output_tokens;
  const total = usage.total_tokens ?? (n(prompt) + n(completion));
  return {
    prompt_tokens: n(prompt),
    completion_tokens: n(completion),
    total_tokens: n(total),
  };
}

function parseUsageRecords(lines) {
  const rows = [];
  for (const line of lines) {
    if (!line.event) continue;
    if (!["usage", "analyzer_usage"].includes(line.event)) continue;
    if (!line.raw || typeof line.raw !== "object") continue;
    const usage = line.raw.usage;
    const tokens = extractTokens(usage);
    if (!tokens) continue;
    rows.push({
      model: line.raw.modelId || "unknown",
      sessionKey: line.raw.sessionKey || "",
      keyId: line.raw.keyId || "",
      ...tokens,
    });
  }
  return rows;
}

function parseTimingRecords(lines) {
  const rows = [];
  for (const line of lines) {
    if (!line.event) continue;
    if (line.event !== "request.timing") continue;
    const d = line.data;
    const total = d.totalMs ?? d.total_ms;
    if (typeof total !== "number") continue;
    rows.push({
      model: d.model || "unknown",
      protocol: d.protocol || "unknown",
      stream: d.stream ?? null,
      totalMs: Math.round(total || 0),
      ttfbMs: Math.round(d.upstreamTtfbMs ?? d.upstream_ttfb_ms ?? 0),
      streamMs: Math.round(d.upstreamStreamMs ?? d.upstream_stream_ms ?? 0),
    });
  }
  return rows;
}

function pct(sorted, q) {
  if (sorted.length === 0) return 0;
  if (sorted.length === 1) return sorted[0];
  const pos = q * (sorted.length - 1);
  const lo = Math.floor(pos);
  const hi = Math.min(lo + 1, sorted.length - 1);
  return Math.round(sorted[lo] + (sorted[hi] - sorted[lo]) * (pos - lo));
}

function printUsage(rows) {
  console.log("\n== Token / Usage ==");
  if (rows.length === 0) {
    console.log("(no usage rows found)");
    return;
  }
  const totalTokens = rows.reduce((s, r) => s + r.total_tokens, 0);
  console.log(`rows=${rows.length} total_tokens=${totalTokens}`);

  const byModel = new Map();
  for (const r of rows) {
    const m = byModel.get(r.model) ?? { rows: 0, total_tokens: 0 };
    m.rows += 1;
    m.total_tokens += r.total_tokens;
    byModel.set(r.model, m);
  }
  const sorted = [...byModel.entries()].sort((a, b) => b[1].rows - a[1].rows);
  console.log("model".padEnd(20), "calls".padStart(6), "tokens".padStart(12));
  for (const [model, m] of sorted) {
    console.log(model.padEnd(20), String(m.rows).padStart(6), String(m.total_tokens).padStart(12));
  }
}

function printTiming(rows) {
  console.log("\n== Timing ==");
  if (rows.length === 0) {
    console.log("(no request.timing rows found)");
    return;
  }
  console.log(`rows=${rows.length}`);

  const groups = new Map();
  for (const r of rows) {
    const key = `${r.model}\u0000${r.protocol}\u0000${r.stream}`;
    const g = groups.get(key) ?? [];
    g.push(r);
    groups.set(key, g);
  }
  console.log(
    "model".padEnd(16),
    "protocol".padEnd(10),
    "stream".padStart(3),
    "requests".padStart(7),
    "p50_total_ms".padStart(13),
    "p90_total_ms".padStart(13),
    "p50_ttfb_ms".padStart(12),
    "p90_ttfb_ms".padStart(12),
  );
  const sorted = [...groups.entries()].sort((a, b) => b[1].length - a[1].length);
  for (const [, group] of sorted) {
    const totals = group.map((r) => r.totalMs).sort((a, b) => a - b);
    const ttfb = group.map((r) => r.ttfbMs).sort((a, b) => a - b);
    const first = group[0];
    console.log(
      String(first.model).padEnd(16),
      String(first.protocol).padEnd(10),
      String(first.stream ?? "").padStart(3),
      String(group.length).padStart(7),
      String(pct(totals, 0.5)).padStart(13),
      String(pct(totals, 0.9)).padStart(13),
      String(pct(ttfb, 0.5)).padStart(12),
      String(pct(ttfb, 0.9)).padStart(12),
    );
  }
}

function main() {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    console.error("Usage: node query_usage_stats.mjs <proxy.log|jsonl|log-dir> [...]");
    process.exit(1);
  }

  const files = collectFiles(args);
  if (files.length === 0) {
    console.error("No log files found in the given paths.");
    process.exit(1);
  }

  const lines = [];
  for (const file of files) {
    const text = fs.readFileSync(file, "utf8");
    for (const raw of text.split("\n")) {
      const { event, data } = parseLine(raw);
      if (!event) continue;
      const rawObj = raw.startsWith("{") ? JSON.parse(raw) : data;
      lines.push({ event, data, raw: event === "request.timing" ? data : rawObj });
    }
  }

  printUsage(parseUsageRecords(lines));
  printTiming(parseTimingRecords(lines));
}

main();
